### PR TITLE
Simplify Chrome Android setup

### PR DESCRIPTION
**Pull request by Hexcles**
Automatically start adb, connect to device, and set up port forwarding.

---
This change is [Reviewable](https://reviewable.io/reviews/w3c/web-platform-tests/9188)

### CODE REVIEW

**Comment by w3c-bots** [comment]
# Build PASSED

Started: 2018-01-25 01:31:08
Finished: 2018-01-25 01:48:43





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/23260)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/333047600)

**Comment by tobie** [comment]
Summoning @wpt-pr-bot.

**Comment by Hexcles** [comment]
@jgraham the PR is ready for review

Most of the change is incremental improvements in our draft Chrome Android setup, except this line, which fixes the setup/cleanup imbalance I told you about:
https://github.com/w3c/web-platform-tests/pull/9188/commits/f71d30a6f878257e05432aceac7c0456857d10d8#diff-35e3a18456dc70eea5ca7fff107a1a46L232

File: tools/wpt/browser.py [hunk 1]
```diff
@@ -35,6 +35,10 @@ def install(self, dest=None):
     def install_webdriver(self):
         return NotImplemented
 
+    @abstractmethod
+    def find_webdriver(self):
```
Comment (foolip) [inline tools/wpt/browser.py]: With no other changes to `find_webdriver` in this PR, an explanation of why this change is needed in the commit message would be good. Does this mean that a subclass not implementing this will blow up? (You can write a block in a comment describing what the commit message should be when merging.)

File: tools/wptrunner/wptrunner/browsers/chrome_android.py [hunk 2]
```diff
@@ -76,6 +79,19 @@ def __init__(self, logger, binary, webdriver_binary="chromedriver",
...
 
+    def _adb_run(self, args):
+        self.logger.info('adb ' + ' '.join(args))
+        subprocess.check_call(['adb'] + args)
+
+    def setup(self):
+        self._adb_run(['wait-for-device'])
+        self._adb_run(['forward', '--remove-all'])
+        self._adb_run(['reverse', '--remove-all'])
+        # TODO: Use ports from config.
```
Comment (foolip) [inline tools/wptrunner/wptrunner/browsers/chrome_android.py]: Some logging confirms that there are in `server_config` in `executor_kwargs`, and I'm guessing the `executor_kwargs` return value ends up somehow propagated to the `ChromeAndroidBrowser` instance? In other words, is this easy to fix already? If so let's fix it, if it's hard because of some mis-layering, maybe file an issue about that layering problem.

**Review by foolip — COMMENTED** [review]
If @jgraham wants to take a look that would be most kind, but I'm happy to do my best and have tried.

I also tried running the `wpt run chrome_android speech-api` on my Mac and did see Chrome on my phone doing some things, but after a few tries I see a lot of this:
```
objc[66566]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called.
objc[66566]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
```

And then nothing seems to really work. But I haven't rooted this phone to update the hosts file, and it seems to work about the same without the changes. If confirmed to actually work somewhere, that's good! :)

File: tools/wpt/browser.py (hunk 1 above)
Comment (Hexcles) [reply tools/wpt/browser.py]: Actually, let me hack the history of this branch. I'll make a feature change commit + a code health commit. And then we can rebase-merge this PR.

File: tools/wptrunner/wptrunner/browsers/chrome_android.py (hunk 2 above)
Comment (Hexcles) [reply tools/wptrunner/wptrunner/browsers/chrome_android.py]: No, it doesn't. Although there is a hook for Browser modules to modify `executor_kwargs`, but the dictionary is not passed back to Browser classes, as Browsers are not Executors... I'm going to store the `server_config` in a module-local variable. It's not ideal, but probably better than hardcoding.

**Comment by Hexcles** [comment]
I ended up splitting the changes into two PRs. See #9349 for the follow-up code health fixes.

**Review by foolip — APPROVED** [review]

File: tools/wptrunner/wptrunner/testrunner.py [hunk 3]
```diff
@@ -229,7 +229,6 @@ def stop(self, force=False):
     def cleanup(self):
         if self.init_timer is not None:
             self.init_timer.cancel()
-        self.browser.cleanup()
```
Comment (foolip) [inline tools/wptrunner/wptrunner/testrunner.py]: Wait was this removal intentional?

Comment (Hexcles) [reply tools/wptrunner/wptrunner/testrunner.py]: Yes, indeed, and it's quite important. See the commit message:

> Also fix testrunner so that setup & cleanup balance out (called at enter
> & exit of a Browser context, respectively). Previously, cleaup() didn't
> do anything in all browsers, so the issue wasn't discovered. In this
> commit, cleanup() is used for uninstalling port forwarding.

which is all about the removal of this one line.

**Comment by Hexcles** [comment]
Also, @foolip regarding the errors you saw, I'm quite certain they are not related to my change. And I'm quite confused: where exactly do we ever call a compiler (objc)?

**Comment by gsnedders** [comment]
@foolip's errors appear to just be #6998 manifesting in a slightly less crashy way.
